### PR TITLE
Flume, Metrics and Message control when overloaded

### PR DESCRIPTION
**Pull request by jsdw**
- **Make use of `flume` throughout**. It seems to benchmark well against the futures channels I've been using (see https://github.com/zesterer/flume/blob/master/misc/benchmarks.png) but importantly it also exposes a `len()` fn so we can see how many messages are queued in each channel.
- **Gather and expose metrics**. The above gives us access to some useful metrics on top of what we'd already have. These are exposed in a format compatible with Prometheus (tested using a local version of prometheus).
- **Test util updates**. Remove the "worse" soak test runner, and add options to make it easier to test a few different things, and easier to scale the test runner to use more cores (which is needed for larger scale tests).

Builds on https://github.com/paritytech/substrate-telemetry/tree/jsdw-minor-fixes. Merge this after that.

### CODE REVIEW

**Comment by jsdw** [comment]
Note to self: change base to master, and rebase/merge this after the `jsdw-minor-fixes` stuff merges

**Comment by jsdw** [comment]
(force push to undo a commit I put on the wrong branch :))

File: backend/common/src/dense_map.rs
Comment (dvdplm) [inline backend/common/src/dense_map.rs:141]: Did you mean "removed" here?

File: backend/common/src/ws_client/receiver.rs
Comment (dvdplm) [inline backend/common/src/ws_client/receiver.rs:23]: Eew, that is an unfortunate choice of module name.

File: backend/telemetry_core/src/aggregator/aggregator.rs
Comment (dvdplm) [inline backend/telemetry_core/src/aggregator/aggregator.rs:89]:
```suggestion
    /// in to the aggregator. If nobody is holding the tx side of the channel
```

File: backend/telemetry_core/src/aggregator/inner_loop.rs
Comment (dvdplm) [inline backend/telemetry_core/src/aggregator/inner_loop.rs:120]:
```suggestion
    /// How many nodes are currently known to this aggregator.
```
?

File: backend/telemetry_core/src/aggregator/inner_loop.rs
Comment (dvdplm) [inline backend/telemetry_core/src/aggregator/inner_loop.rs:123]: dq: what is the difference between connected feeds and subscribed feeds?

File: backend/telemetry_core/src/aggregator/inner_loop.rs
Comment (dvdplm) [inline backend/telemetry_core/src/aggregator/inner_loop.rs:106]: Thinking out loud here: it would be interesting to have a few histograms too, e.g. nodes and their verbosity level (to get an inkling to how much data we're receiving), message payload size in and out, maybe connection longevity too (how often do nodes stay connected).
Another data point I wouldn't mind having is the current message rate (is that what they call a "gauge"?).

File: backend/telemetry_core/src/main.rs
Comment (dvdplm) [inline backend/telemetry_core/src/main.rs:490]:
```suggestion
    // we just split out the text format that prometheus expects ourselves, and use the latest metrics that we've
    // captured so far from the aggregators. See:
```

**Review by dvdplm — APPROVED** [review]
LGTM. Left a few nits and suggestions (likely for future work).

File: backend/telemetry_core/src/aggregator/aggregator_set.rs
Comment (insipx) [inline backend/telemetry_core/src/aggregator/aggregator_set.rs:74]: why `sleep_until` instead of just `sleep`?

File: backend/telemetry_core/src/aggregator/inner_loop.rs
Comment (insipx) [inline backend/telemetry_core/src/aggregator/inner_loop.rs:41]:
```suggestion
    /// a message is sent into it.
```

**Review by insipx — APPROVED** [review]

File: backend/common/src/ws_client/receiver.rs
Comment (jsdw) [reply backend/common/src/ws_client/receiver.rs:23]: Yeah, I thought so! If I had to write that too much I'd probably `use` and rename it!

File: backend/common/src/dense_map.rs
Comment (jsdw) [reply backend/common/src/dense_map.rs:141]: `retired` is referring to the list of `retired` IDs in the densemap struct :)

File: backend/telemetry_core/src/aggregator/inner_loop.rs
Comment (jsdw) [reply backend/telemetry_core/src/aggregator/inner_loop.rs:113]: What "too" was hinting at is that it's the number of feeds that are subscribed to a chain and also have asked for finality (because you can't ask for finality without being subscribed)

File: backend/telemetry_core/src/aggregator/inner_loop.rs
Comment (jsdw) [reply backend/telemetry_core/src/aggregator/inner_loop.rs:123]: As soon as a connection to `/feed` is established, the aggregator can see that as a connected feed. Only when the feed sends a message like "subscribe:Polkadot" does it also count as having subscribed to something!

File: backend/telemetry_core/src/aggregator/inner_loop.rs
Comment (jsdw) [reply backend/telemetry_core/src/aggregator/inner_loop.rs:106]: I think I return the current message rates already (I think guage is the right term.. there was a word for numbers that are always increasing and a word for values that can go up or down and I think guage was the latter!). I think we can see the bandwidth anyway from outside the process, so I wonder how useful the bytes in/out are? But that histogram sounds like a good idea; roughly breaking down the bytes in/out per node/feed could be interesting!

When this merges, perhaps we should create an issue to track any additional metrics we'd like to add?

File: backend/telemetry_core/src/aggregator/aggregator_set.rs
Comment (jsdw) [reply backend/telemetry_core/src/aggregator/aggregator_set.rs:74]: I could have gone either way really, but I went for "aim to have 10 seconds between each update to metrics, unless it takes more than 10 seconds for an update to come through"

**Comment by jsdw** [comment]
> looks good to me apart from a couple questions. Overall I've had a good experience with flume, I like how Sync/Async methods are both available on the same Sender/Receiver objects

That's good to hear! I like the flexibility it has there too, and unlike the tokio channels it also has Stream/Sink impls, and unlike futures/tokio channels you can see lengths (and on top of all of that, it's supposed to be pretty quick) :)

File: backend/telemetry_core/src/aggregator/aggregator_set.rs
Comment (jsdw) [reply backend/telemetry_core/src/aggregator/aggregator_set.rs:74]: (To explain the reasoning a bit more; the actual logic to gather metrics is fairly quick, but potentially the message could end up in a queue for a while if we happen to be under heavy load, so if it takes eg 9seconds to get a response, I'd rather put another message into the queue 1 second later than wait another 10 seconds before queueing another message)

File: backend/telemetry_core/src/aggregator/inner_loop.rs
Comment (jsdw) [reply backend/telemetry_core/src/aggregator/inner_loop.rs:240]: Ooh yeah, that's a really good number to have! I might sneak that into this PR as I don't think it'll be that hard to gather and it would be useful to see reported!

File: backend/telemetry_core/src/aggregator/inner_loop.rs
Comment (dvdplm) [reply backend/telemetry_core/src/aggregator/inner_loop.rs:106]: > When this merges, perhaps we should create an issue to track any additional metrics we'd like to add?

You read my mind.

File: backend/common/src/dense_map.rs
Comment (jsdw) [reply backend/common/src/dense_map.rs]: (I made the name less awful!)

File: backend/telemetry_core/src/aggregator/inner_loop.rs
Comment (dvdplm) [reply backend/telemetry_core/src/aggregator/inner_loop.rs:114]: I see. How about this: "Number of subscribed feeds that also asked for finality information."?

File: backend/telemetry_core/src/aggregator/inner_loop.rs
Comment (dvdplm) [reply backend/telemetry_core/src/aggregator/inner_loop.rs:129]: So one connection could subscribe to multiple chains?

File: backend/telemetry_core/src/aggregator/inner_loop.rs
Comment (dvdplm) [reply backend/telemetry_core/src/aggregator/inner_loop.rs:255]: Please do!

Comment (jsdw) [reply backend/telemetry_core/src/aggregator/inner_loop.rs:255]: (done!)

File: backend/telemetry_core/src/aggregator/inner_loop.rs
Comment (jsdw) [reply backend/telemetry_core/src/aggregator/inner_loop.rs:129]: Nope; if a connection calls "susbcribe:ChainName" it'll unsubscribe them from any previous chain they were subscribed to, so it's just one at a time.

File: backend/telemetry_core/src/aggregator/inner_loop.rs
Comment (jsdw) [reply backend/telemetry_core/src/aggregator/inner_loop.rs:117]: I like it!

File: backend/common/src/dense_map.rs
Comment (dvdplm) [inline backend/common/src/dense_map.rs:141]: ❤️

File: backend/telemetry_core/src/aggregator/inner_loop.rs
Comment (dvdplm) [reply backend/telemetry_core/src/aggregator/inner_loop.rs:129]: Got it.

**Comment by dvdplm** [comment]
Last [changes](https://github.com/paritytech/substrate-telemetry/pull/367/commits/b842c7fc8bbfa5b25cd087283efbd6a62db55790) still lgtm! Merge at will.

**Pull request merged**
